### PR TITLE
Add batching and checkpoint to energy fun map enable memory management

### PR DESCRIPTION
**Pull request by amitschang**
This PR replaces the `vmap` in energy function map with `jax.lax.map` with a configurable batch_size and adds checkpointing if configured. This enables control over the peak memory usage during objective gradient computation.

For the martini case, `vmap` was causing OOM on a 500GB system when run against 10k states, whereas with the `map` and `batch_size=100` the mem is within ~3GB. Using checkpoint enables mem constraint for the `grad` part, in the same martini case using checkpoint and same batch size brings down from 250GB to ~5GB.

Good values are problem dependent, but some reasonable starting values providing a decent tradeoff are included. These could come from some env variables as well?

### CODE REVIEW

**Review by ryanhausen — APPROVED** [review]
LGTM, nice fix.
